### PR TITLE
[wip] prevent user overwrite default configuration 

### LINKED ISSUE #421: restrict some fields from being set in default values
#419 implements defaulting logic in decK.
A user can set all properties of Service/Route/Target/Upstream but not all of them actually make sense.
The goal of this issue is to add validation and error out if the following fields are specified in data structure KongDefaults (Under Info of Content):
- service.id
- service.host
- service.name
- service.port
- route.name
- route.id
- target.target4
- target.id
- upstream.id
- upstream.name

### PR DESCRIPTION
**Pull request by ccfishk**
prevent user overwrite some default configuration.

fixes #421

### CODE REVIEW

**Comment by hbagdi** [comment]
feedback from https://github.com/Kong/deck/pull/441 hasn't yet been addressed. Please add tests and resolve the nil reference issues.

**Comment by ccfishk** [comment]
> feedback from #441 hasn't yet been addressed. Please add tests and resolve the nil reference issues.

Tests are added, and nil reference was removed if I understand correctly.

File: file/validate.go [hunk 1]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+	}
+	if kongdefaults != nil {
+		if err := CheckDefaults(*kongdefaults); err != nil {
+			return fmt.Errorf("default values are not allowed for these fields: %w", err)
+		}
+	}
 	return errs
 }
+
+func Check(item string, t *string) string {
```
Comment (hbagdi) [inline file/validate.go:47]: Please lowercase this to not export this function.

File: file/validate.go [hunk 2]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+	if t != nil && len(*t) > 0 {
+		return item
+	}
+	return ""
+}
+
+func CheckDefaults(kd KongDefaults) error {
```
Comment (hbagdi) [inline file/validate.go:54]: Please lowercase this to not export this function.

File: file/validate.go [hunk 3]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+	var invalid []string
+	if ret := Check("Service.ID", kd.Service.ID); len(ret) > 0 {
```
Comment (hbagdi) [inline file/validate.go:56]: kd.Service is a pointer. What happens if it is a `nil`?

File: file/validate.go [hunk 4]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+		invalid = append(invalid, "Service.ID")
+	}
+
+	if ret := Check("Service.Host", kd.Service.Host); len(ret) > 0 {
```
Comment (hbagdi) [inline file/validate.go:60]: kd.Service is a pointer. What happens if it is a `nil`?

File: file/validate.go [hunk 5]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+		invalid = append(invalid, "Service.Host")
+	}
+
+	if ret := Check("Service.Name", kd.Service.Name); len(ret) > 0 {
```
Comment (hbagdi) [inline file/validate.go:64]: kd.Service is a pointer. What happens if it is a `nil`?

File: file/validate.go [hunk 6]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+		invalid = append(invalid, "Service.Name")
+	}
+
+	if kd.Service.Port != nil {
```
Comment (hbagdi) [inline file/validate.go:68]: kd.Service is a pointer. What happens if it is a `nil`?

File: file/validate.go [hunk 7]
```diff
@@ -29,5 +31,70 @@ func validate(content []byte) error {
...
+		invalid = append(invalid, "Service.Port")
+	}
+
+	if ret := Check("Route.Name", kd.Route.Name); len(ret) > 0 {
```
Comment (hbagdi) [inline file/validate.go:72]: Same comments for most of these lines..

**Review by hbagdi — CHANGES_REQUESTED** [review]

File: file/validate.go (hunk 3 above)
Comment (ccfishk) [reply file/validate.go:56]: hmm, possibly, if deck has another client other than KIC.

File: file/validate.go (hunk 1 above)
Comment (ccfishk) [reply file/validate.go]: changed 67aa05d6a0a1841031b4910928070b31bfaf9be8

File: file/validate.go (hunk 3 above)
Comment (ccfishk) [reply file/validate.go]: changed 67aa05d6a0a1841031b4910928070b31bfaf9be8

File: file/validate.go (hunk 5 above)
Comment (ccfishk) [reply file/validate.go]: checked 67aa05d6a0a1841031b4910928070b31bfaf9be8

File: file/validate.go (hunk 2 above)
Comment (ccfishk) [reply file/validate.go]: changed 67aa05d6a0a1841031b4910928070b31bfaf9be8

File: file/validate.go (hunk 4 above)
Comment (ccfishk) [reply file/validate.go]: checked 67aa05d6a0a1841031b4910928070b31bfaf9be8

File: file/validate.go (hunk 7 above)
Comment (ccfishk) [reply file/validate.go]: checked 67aa05d6a0a1841031b4910928070b31bfaf9be8

File: file/validate.go [hunk 8]
```diff
@@ -29,5 +31,86 @@ func validate(content []byte) error {
 		err := fmt.Errorf(desc.String())
 		errs.Errors = append(errs.Errors, err)
 	}
+
+	var kongdefaults *KongDefaults
+	if err := json.Unmarshal(content, kongdefaults); err != nil {
```
Comment (hbagdi) [inline file/validate.go:36]: Have you actually tried running decK with this patch?
This won't work at all. The content in question should represents the following struct: https://github.com/Kong/deck/blob/main/file/types.go#L597-L617

File: file/validate.go [hunk 9]
```diff
@@ -29,5 +31,86 @@ func validate(content []byte) error {
...
+	if t != nil && len(*t) > 0 {
+		return item
+	}
+	return ""
+}
+
+func checkDefaults(kd KongDefaults) error {
+	var invalid []string
+	if kd.Service == nil {
+		invalid = append(invalid, "Invalid Service Address")
```
Comment (hbagdi) [inline file/validate.go:57]: This is not an error because the user can choose to not specify a default for service.
If service is nil, you can skip checking any of the properties all together and that's it.

The same comment applies for route, target and upstream.

File: file/validate.go (hunk 8 above)
Comment (ccfishk) [reply file/validate.go]: Yesterday's offline discussion KIC is a deck client, and the File.Content is the main construct I explored atm (edited) 
white_check_mark
eyes
raised_hands





10:04
Assuming decK applies the same interface to all clients.

File: file/validate.go (hunk 9 above)
Comment (ccfishk) [reply file/validate.go]: changed 507639081e5cb99ce9e720a28e58c17b7afdafc4
